From 6167b7ae9d2c965c9f0edfc925456d417e5de954 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Thu, 30 Apr 2015 12:24:06 +0300
Subject: [PATCH 087/214] dpaa2-eth: Fix "NOHZ: local_softirq_pending" warning

Explicitly run softirqs after we enable NAPI. This in particular gets us
rid of the "NOHZ: local_softirq_pending" warnings, but it also solves a
couple of other problems, among which fluctuating performance and high
ping latencies.

Notes:
 - This will prevent us from timely processing notifications and
other "non-frame events" coming into the software portal. So far,
though, we only expect Dequeue Available Notifications, so this patch
is good enough for now.
 - A degradation in console responsiveness is expected, especially in
cases where the bottom-half runs on the same CPU as the console.

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Signed-off-by: Ioana Radulescu <ruxandra.radulescu@freescale.com>
Change-Id: Ia6f11da433024e80ee59e821c9eabfa5068df5e5
Reviewed-on: http://git.am.freescale.net:8181/35830
Reviewed-by: Alexandru Marginean <Alexandru.Marginean@freescale.com>
Reviewed-by: Stuart Yoder <stuart.yoder@freescale.com>
Tested-by: Stuart Yoder <stuart.yoder@freescale.com>
[Original patch extracted from FSL LS2085 SDK EAR5.0,
Layerscape2-SDK-SOURCE-20150828-yocto.iso]
Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c b/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
index a8ca95b..933b4e9 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
@@ -1122,6 +1122,16 @@ static void ldpaa_eth_fqdan_cb(struct dpaa_io_notification_ctx *ctx)
 
 	fq->has_frames = true;
 	napi_schedule(&fq->napi);
+	/* Provide a guaranteed scheduling point for the bottom-half;
+	 * with threaded interrupts, that isn't automatically the case.
+	 * FIXME: we're effectively running in the software portal's top-half.
+	 * As long as:
+	 *   1. the Ethernet driver is the only client of the portal, and
+	 *   2. we only expect Dequeue Available Notifications,
+	 * this approach is fine. Once either of the conditions no longer holds,
+	 * we will have to move this to a separate execution context.
+	 */
+	do_softirq();
 }
 
 static void ldpaa_eth_setup_fqs(struct ldpaa_eth_priv *priv)
-- 
2.5.3

